lib/internal/modules/esm/hooks.js

```diff
@@ -516,9 +460,187 @@ class Hooks {
     };
   }
 }
-
 ObjectSetPrototypeOf(Hooks.prototype, null);
 
+let MessageChannel;
+class HooksProxy {
+  /**
+   * The lock/unlock segment of the shared memory. Atomics require this to be a Int32Array. This
+   * segment is used to tell the main to sleep when the worker is processing.
+   * 0 -> main sleeps
+   * 1 -> main wakes up
+   */
+  #lock;
+  /**
+   * A MessagePort used to synchronously communicate with the worker.
+   */
+  #syncCommPort;
+
+  #isReady = false;
+
+  constructor() {
+    const { InternalWorker } = require('internal/worker');
+    MessageChannel ??= require('internal/worker/io').MessageChannel;
+
+    const lock = new SharedArrayBuffer(SHARED_MEMORY_BYTE_LENGTH);
+    this.#lock = new Int32Array(lock);
+    const syncCommChannel = new MessageChannel();
+    this.#syncCommPort = syncCommChannel.port1;
+
+    const worker = this.worker = new InternalWorker('internal/modules/esm/worker', {
+      stderr: false,
+      stdin: false,
+      stdout: false,
+      trackUnmanagedFds: false,
+      workerData: {
+        lock,
+        syncCommPort: syncCommChannel.port2,
+      },
+      transferList: [syncCommChannel.port2],
+    });
+    worker.unref(); // ! Allows the process to eventually exit.
+  }
+
+  #waitForWorker() {
+    if (!this.#isReady) {
+      const { kIsOnline } = require('internal/worker');
+      if (!this.worker[kIsOnline]) {
+        debug('wait for signal from worker');
+        AtomicsWait(this.#lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 0);
+        const response = receiveMessageOnPort(this.#syncCommPort);
+        const { preloadScripts } = this.#unwrapMessage(response);
+        this.#executePreloadScripts(preloadScripts);
+      }
+
+      this.#isReady = true;
+    }
+  }
+
+  #beforeExitHandler = () => {
+    const value = AtomicsOr(this.#lock, IDLENESS_OF_THREADS, MAIN_THREAD_IS_IDLE);
+    debug('beforeExit main thread', value, this.#lock);
+    if (value & WORKER_THREAD_IS_IDLE && AtomicsLoad(this.#lock, NUMBER_OF_INCOMING_MESSAGES) === 0) {
+      debug('send terminate order to worker');
+      // The worker has already reported waiting for us to die together.
+      this.worker.terminate();
+    } else {
+      // The worker is still in good health, and we just told it it's the
+      // only thing keeping us alive.
+      this.worker.ref();
+    }
+  };
+
+  async makeAsyncRequest(method, ...args) {
+    AtomicsAdd(this.#lock, NUMBER_OF_INCOMING_MESSAGES, 1);
+    this.#waitForWorker();
```

Comment:
Does this now support concurrency?

Reply:
This stops the main thread until the worker thread is ready, there’s no concurrency possible if the thread is locked.

Reply:
Why do we need `NUMBER_OF_INCOMING_MESSAGES` then if they're resolved 1 by 1?

Reply:
There can be several messages sent before any is handled, in practice it's going to be mostly 0 or 1, but it could be more.
I think I don't understand what "it" refers to in your original question, could you rephrase maybe?

Reply:
That answers my question, thanks 🙂